### PR TITLE
Feature: Manage ACL for Jazz services

### PR DESCRIPTION
**Pull request by devsatishm**
### Feature: 

Current version of Jazz works in a single developer mode where the service creator/owner has full access to their service with no provision to share access within the team. This feature will expand the scope of using Jazz in a collaborative team/application/product setup.

### Benefits:

Service owners will be able to add/remove users to specific roles for their services. Supported access categories & roles: 

| Category        | Role           | Description  |
| ------------- |:-------------:| -----|
| Manage | admin | Administer the service in Jazz UI. Users in this role will have ALL permissions on this service.|
| Manage | read | Read-only permissions on the service in Jazz UI |
| Code | write |  Users with this role get read-write permissions to the code |
| Code | read |  Users with this role get read-only permissions to the code |
| Deploy | write |  Users with this role get access to trigger deployments to this service in Jazz UI |
| Deploy | read |   Users with this role get read-only access to deployment logs for this service |

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,6 +2,7 @@ language: node_js
 node_js:
  - "node"
 env:
+ - PLATFORM_SERVICE=core/jazz_acl
```
Comment (devsatishm) [inline .travis.yml]: Add the authorizers to this list if unit tests are aded.

File: core/jazz_acl/components/validation.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+// limitations under the License.
+// =========================================================================
+
+const errorHandlerModule = require("./error-handler.js")();
+
+function validateBasicInput(userInput) {
+    if (!userInput) {
+      throw (errorHandlerModule.throwInternalServerError("Input parameters are missing"));
+    }
+
```
Comment (devsatishm) [inline core/jazz_acl/components/validation.js]: Fix the formatting.

File: core/jazz_acl/config/rbac_model.conf [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+p = sub, obj, act
+
+[role_definition]
+g = _, _
+
+[policy_effect]
+e = some(where (p.eft == allow))
+
+[matchers]
+m=g(r.sub, p.sub) && keyMatch(r.obj, p.obj) && (r.act == p.act || (p.act == "write" || p.act == "admin"))
```
Comment (devsatishm) [inline core/jazz_acl/config/rbac_model.conf]: What does this justify - `(r.act == p.act || (p.act == "write" || p.act == "admin"))`? A note/comment might help (if adding a comment is feasible here)

File: core/jazz_codeq/config/dev-config.json [hunk 4]
```diff
@@ -13,7 +13,8 @@
 		"lines-of-code": "ncloc",
 		"files": "files",
 		"vulnerabilities": "high_severity_vulns"
-	},
+  },
+  "SERVICE_ID_HEADER_KEY": "Jazz-Service-ID",
```
Comment (devsatishm) [inline core/jazz_codeq/config/dev-config.json]: Fix formatting.

File: core/jazz_codeq/index.js [hunk 5]
```diff
@@ -30,15 +30,22 @@ function handler(event, context, cb) {
...
 		logger.debug(serviceContext.resourcePath);
 		utils.getAPIPath(serviceContext.resourcePath)
 			.then(data => {
-				const pathString = data.pathString;
+        const pathString = data.pathString;
+        let serviceId
 				logger.debug(pathString);
 				if (serviceContext && serviceContext.method && serviceContext.method === 'GET' && pathString === "codeq") {
 					logger.debug(`code quality service called with pathstring - ${pathString}`);
+          let header_key = config.SERVICE_ID_HEADER_KEY.toLowerCase();
```
Comment (devsatishm) [inline core/jazz_codeq/index.js]: Fix formatting

File: core/jazz_codeq/index.js [hunk 6]
```diff
@@ -30,15 +30,22 @@ function handler(event, context, cb) {
...
+          if (!serviceContext.headers[header_key]) {
```
Comment (devsatishm) [inline core/jazz_codeq/index.js]: Can you check if other core services (that require serviceId) work in a similar fashion? Check if key exists in header..

File: core/jazz_codeq/swagger/swagger.json [hunk 7]
```diff
@@ -121,6 +121,12 @@
                     "required": false,
                     "type": "number",
                     "description": "page number of the result set."
+                },
```
Comment (devsatishm) [inline core/jazz_codeq/swagger/swagger.json]: Fix formatting across the board.

File: core/jazz_acl/components/validation.js (hunk 2 above)
Comment (pspani) [reply core/jazz_acl/components/validation.js]: fixed

File: core/jazz_acl/config/rbac_model.conf (hunk 3 above)
Comment (pspani) [reply core/jazz_acl/config/rbac_model.conf]: if the user and service_category matches with the request and the user has write/admin permission then they automatically get read permission too. Also added as comment to the code

File: core/jazz_codeq/config/dev-config.json (hunk 4 above)
Comment (pspani) [reply core/jazz_codeq/config/dev-config.json]: fixed

File: core/jazz_codeq/index.js (hunk 5 above)
Comment (pspani) [reply core/jazz_codeq/index.js]: fixed

File: core/jazz_codeq/swagger/swagger.json (hunk 7 above)
Comment (pspani) [reply core/jazz_codeq/swagger/swagger.json]: fixed

File: .travis.yml (hunk 1 above)
Comment (pspani) [reply .travis.yml]: authorizer unit test needs more work

File: builds/jazz-build-module/aws-apigateway-module.groovy [hunk 8]
```diff
@@ -1,49 +1,78 @@
...
+      for (method in methods_of_each_path) {
+        def temp = parsed_json.paths[key_of_a_path][method]
+        if (!temp["x-amazon-apigateway-integration"]) {
+          echo "x-amazon-apigateway-integration does not exist."
+          injectLambdaIntegration(method, filePath);
         }
+      }
     }
+  } catch (ex) {
+    echo " RemoveUserDefinedIntegrationSpec :::: Error occurred" + ex.getMessage()
```
Comment (devsatishm) [inline builds/jazz-build-module/aws-apigateway-module.groovy]: Fix the log message  - RemoveUserDefinedIntegrationSpec

**Pull request merged**